File: js/test/auth_token_manager.test.ts
import { describe, test, expect, vi, beforeEach } from "vitest";
import jwt from "jsonwebtoken";
import { ClientError, Status } from "nice-grpc";
import { ModalClient } from "../src/client";
import {
  AUTH_TOKEN_GET_TIMEOUT_MS,
  AUTH_TOKEN_GET_MAX_RETRIES,
  AuthTokenManager,
  FAILURE_BACKOFF_BASE_MS,
  FAILURE_BACKOFF_MAX_MS,
  REFRESH_WINDOW,
} from "../src/auth_token_manager";
import { newLogger } from "../src/logger";

class mockAuthClient {
  private authToken: string = "";

  setAuthToken(token: string) {
    this.authToken = token;
  }

  authTokenGet = vi.fn(async (_request?: unknown, _options?: unknown) => {
    return { token: this.authToken };
  });
}

function newMockAuthClient() {
  return new mockAuthClient();
}

// Creates a JWT token for testing
function createTestJWT(expiry: number): string {
  return jwt.sign({ exp: expiry }, "walter-test");
}

describe("AuthTokenManager", () => {
  let mockClient: mockAuthClient;
  let manager: AuthTokenManager;

  beforeEach(() => {
    mockClient = newMockAuthClient();
    manager = new AuthTokenManager(mockClient as any, newLogger());
  });

  test("TestAuthToken_DecodeJWT", async () => {
    const now = Math.floor(Date.now() / 1000);
    const expiry = now + 1800;
    const token = createTestJWT(expiry);
    mockClient.setAuthToken(token);

    const result = await manager.getToken();
    expect(result).toBe(token);
    expect(manager.getCurrentToken()).toBe(token);
  });

  test("TestAuthToken_LazyFetch", async () => {
    const now = Math.floor(Date.now() / 1000);
    const token = createTestJWT(now + 3600);
    mockClient.setAuthToken(token);

    // First getToken lazily fetches
    const firstToken = await manager.getToken();
    expect(firstToken).toBe(token);

    // Second getToken returns cached
    const secondToken = await manager.getToken();
    expect(secondToken).toBe(token);

    expect(mockClient.authTokenGet).toHaveBeenCalledTimes(1);
  });

  test("TestAuthToken_FetchOnlyRetriesWithoutCachedToken", async () => {
    const token = createTestJWT(Math.floor(Date.now() / 1000) + 3600);
    mockClient.setAuthToken(token);

    await expect(manager.getToken()).resolves.toBe(token);
    expect(mockClient.authTokenGet.mock.calls[0][1]).toEqual({
      retries: AUTH_TOKEN_GET_MAX_RETRIES,
      timeoutMs: AUTH_TOKEN_GET_TIMEOUT_MS,
    });

    const expiredToken = createTestJWT(Math.floor(Date.now() / 1000) - 60);
    manager.setToken(expiredToken, Math.floor(Date.now() / 1000) - 60);
    await expect(manager.getToken()).resolves.toBe(token);
    expect(mockClient.authTokenGet.mock.calls[1][1]).toEqual({
      retries: 0,
      timeoutMs: AUTH_TOKEN_GET_TIMEOUT_MS,
    });
  });

  test("TestAuthToken_IsExpired", async () => {
    const now = Math.floor(Date.now() / 1000);

    // Test not expired
    const validToken = createTestJWT(now + 3600);
    manager.setToken(validToken, now + 3600);
    expect(manager.isExpired()).toBe(false);

    // Test expired
    const expiredToken = createTestJWT(now - 60);
    manager.setToken(expiredToken, now - 60);
    expect(manager.isExpired()).toBe(true);
  });

  test("TestAuthToken_RefreshExpiredToken", async () => {
    const now = Math.floor(Date.now() / 1000);
    const expiringToken = createTestJWT(now - 60);
    const freshToken = createTestJWT(now + 3600);

    manager.setToken(expiringToken, now - 60);
    mockClient.setAuthToken(freshToken);

    // getToken should see the expired token and fetch a new one
    const token = await manager.getToken();
    expect(token).toBe(freshToken);
  });

  test("TestAuthToken_RefreshNearExpiryToken", async () => {
    const now = Math.floor(Date.now() / 1000);
    // Token within REFRESH_WINDOW of expiry (60s left, window is 300s)
    const expiringToken = createTestJWT(now + 60);
    const freshToken = createTestJWT(now + 3600);

    manager.setToken(expiringToken, now + 60);
    mockClient.setAuthToken(freshToken);

    // getToken should proactively refresh
    const token = await manager.getToken();
    expect(token).toBe(freshToken);
  });

  test("TestAuthToken_ConcurrentGetToken", async () => {
    const token = createTestJWT(Math.floor(Date.now() / 1000) + 3600);
    mockClient.setAuthToken(token);

    // Multiple concurrent getToken calls should all return the same token
    const [result1, result2, result3] = await Promise.all([
      manager.getToken(),
      manager.getToken(),
      manager.getToken(),
    ]);
    expect(result1).toBe(token);
    expect(result2).toBe(token);
    expect(result3).toBe(token);

    // Only one fetch should have happened
    expect(mockClient.authTokenGet).toHaveBeenCalledTimes(1);
  });

  test("TestAuthToken_ConcurrentGetTokenWithExpiredToken", async () => {
    const now = Math.floor(Date.now() / 1000);

    const expiredToken = createTestJWT(now - 10);
    manager.setToken(expiredToken, now - 10);

    const freshToken = createTestJWT(now + 3600);
    mockClient.setAuthToken(freshToken);

    const [result1, result2, result3] = await Promise.all([
      manager.getToken(),
      manager.getToken(),
      manager.getToken(),
    ]);

    expect(result1).toBe(freshToken);
    expect(result2).toBe(freshToken);
    expect(result3).toBe(freshToken);
    expect(mockClient.authTokenGet).toHaveBeenCalledTimes(1);
  });

  test("TestAuthToken_ProactiveRefreshFailureReturnsOldToken", async () => {
    const now = Math.floor(Date.now() / 1000);
    // Token within REFRESH_WINDOW of expiry (60s left, window is 300s)
    const nearExpiryToken = createTestJWT(now + 60);
    manager.setToken(nearExpiryToken, now + 60);

    // Make the refresh RPC fail
    mockClient.authTokenGet.mockRejectedValueOnce(new Error("server blip"));

    // getToken should return the old valid token, not throw
    const token = await manager.getToken();
    expect(token).toBe(nearExpiryToken);
    expect(mockClient.authTokenGet).toHaveBeenCalledTimes(1);
  });

  test("TestAuthToken_NoCachedTokenRefreshFailureRejects", async () => {
    mockClient.authTokenGet.mockRejectedValue(new Error("server blip"));

    await expect(manager.getToken()).rejects.toThrow("server blip");
  });

  test("TestAuthToken_ExpiredRefreshBackoffClearsRefreshPromise", async () => {
    vi.useFakeTimers();
    try {
      const baseTime = new Date("2025-01-01T00:00:00Z");
      vi.setSystemTime(baseTime);
      const now = Math.floor(baseTime.getTime() / 1000);
      const expiredToken = createTestJWT(now - 60);
      const refreshedToken = createTestJWT(now + 3600);
      manager.setToken(expiredToken, now - 60);
      mockClient.authTokenGet.mockRejectedValueOnce(new Error("server blip"));

      await expect(manager.getToken()).resolves.toBe(expiredToken);
      await expect(manager.getToken()).resolves.toBe(expiredToken);
      expect(mockClient.authTokenGet).toHaveBeenCalledTimes(1);

      mockClient.setAuthToken(refreshedToken);
      vi.advanceTimersByTime(FAILURE_BACKOFF_BASE_MS + 1);

      await expect(manager.getToken()).resolves.toBe(refreshedToken);
      expect(mockClient.authTokenGet).toHaveBeenCalledTimes(2);
    } finally {
      vi.useRealTimers();
    }
  });

  test("TestAuthToken_GetToken_EmptyResponse", async () => {
    // authToken is "" by default, so authTokenGet returns empty
    await expect(manager.getToken()).rejects.toThrow(
      "did not receive auth token from server",
    );
  });

  test("TestAuthToken_RefreshBackoffGrowsExponentially", async () => {
    vi.useFakeTimers();
    try {
      const baseTime = new Date("2025-01-01T00:00:00Z");
      vi.setSystemTime(baseTime);
      const now = Math.floor(baseTime.getTime() / 1000);
      const expiredToken = createTestJWT(now - 60);
      const freshToken = createTestJWT(now + 3600);
      manager.setToken(expiredToken, now - 60);
      mockClient.authTokenGet.mockRejectedValue(new Error("server blip"));

      const expectedBackoffs = [
        FAILURE_BACKOFF_BASE_MS,
        1000,
        2000,
        4000,
        8000,
        16000,
        32000,
        FAILURE_BACKOFF_MAX_MS,
        FAILURE_BACKOFF_MAX_MS,
      ];
      for (const expectedBackoff of expectedBackoffs) {
        await expect(manager.getToken()).resolves.toBe(expiredToken);
        expect(manager["retryAfter"]).toBe(Date.now() + expectedBackoff);
        vi.setSystemTime(new Date(Date.now() + expectedBackoff + 1));
      }

      mockClient.setAuthToken(freshToken);
      mockClient.authTokenGet.mockResolvedValue({ token: freshToken });
      await expect(manager.getToken()).resolves.toBe(freshToken);
      expect(manager["backoffMs"]).toBe(FAILURE_BACKOFF_BASE_MS);

      manager.setToken(expiredToken, now - 60);
      mockClient.authTokenGet.mockRejectedValue(new Error("server blip"));
      await expect(manager.getToken()).resolves.toBe(expiredToken);
      expect(manager["retryAfter"]).toBe(Date.now() + FAILURE_BACKOFF_BASE_MS);
    } finally {
      vi.useRealTimers();
    }
  });

  test.each([Status.UNAUTHENTICATED, Status.PERMISSION_DENIED])(
    "TestAuthToken_AuthDeniedDoesNotFallBack (%s)",
    async (status) => {
      const now = Math.floor(Date.now() / 1000);
      const expiredToken = createTestJWT(now - 60);
      manager.setToken(expiredToken, now - 60);
      mockClient.authTokenGet.mockRejectedValue(
        new ClientError("/auth-token", status, "credentials rejected"),
      );

      await expect(manager.getToken()).rejects.toThrow("credentials rejected");
      expect(manager.getCurrentToken()).toBe(expiredToken);
      expect(manager["backoffMs"]).toBe(FAILURE_BACKOFF_BASE_MS);
      expect(manager["inBackoff"]()).toBe(false);
    },
  );

  test("TestAuthToken_TransientFailureFallsBackAndBacksOff", async () => {
    const now = Math.floor(Date.now() / 1000);
    const expiredToken = createTestJWT(now - 60);
    manager.setToken(expiredToken, now - 60);
    mockClient.authTokenGet.mockRejectedValue(
      new ClientError("/auth-token", Status.UNAVAILABLE, "server unavailable"),
    );

    await expect(manager.getToken()).resolves.toBe(expiredToken);
    expect(manager["backoffMs"]).toBe(2 * FAILURE_BACKOFF_BASE_MS);
    expect(manager["inBackoff"]()).toBe(true);
    expect(manager["retryAfter"]).toBeGreaterThan(Date.now());
  });

  test("TestAuthToken_ExpiredThenRefreshed", async () => {
    vi.useFakeTimers();
    try {
      const baseTime = new Date("2025-01-01T00:00:00Z");
      vi.setSystemTime(baseTime);
      const baseTimeSeconds = Math.floor(baseTime.getTime() / 1000);

      const tokenOneExpirySeconds = baseTimeSeconds + REFRESH_WINDOW + 5;

      // First getToken lazily fetches tokenOne
      const tokenOne = createTestJWT(tokenOneExpirySeconds);
      mockClient.setAuthToken(tokenOne);
      await expect(manager.getToken()).resolves.toBe(tokenOne);
      expect(mockClient.authTokenGet).toHaveBeenCalledTimes(1);

      // Simulate time moving past token expiry
      const tokenTwo = createTestJWT(tokenOneExpirySeconds + 3600);
      mockClient.setAuthToken(tokenTwo);
      vi.setSystemTime(new Date((tokenOneExpirySeconds + 1) * 1000));

      // getToken should see tokenOne expired and fetch tokenTwo
      await expect(manager.getToken()).resolves.toBe(tokenTwo);
      expect(mockClient.authTokenGet).toHaveBeenCalledTimes(2);
    } finally {
      vi.useRealTimers();
    }
  });
});

describe("ModalClient with AuthTokenManager", () => {
  test("TestModalClient_CloseCleansUpAuthTokenManager", () => {
    const mockCpClient = newMockAuthClient();
    const client = new ModalClient({
      cpClient: mockCpClient as any,
    });

    client.close();
  });

  test("TestModalClient_MultipleInstancesHaveSeparateManagers", () => {
    const mockCpClient1 = newMockAuthClient();
    const mockCpClient2 = newMockAuthClient();

    const client1 = new ModalClient({
      cpClient: mockCpClient1 as any,
    });

    const client2 = new ModalClient({
      cpClient: mockCpClient2 as any,
    });

    client1.close();
    client2.close();
  });
});

File: py/test/auth_token_manager_test.py
# Copyright Modal Labs 2024
import asyncio
import pytest
import time

import jwt

from modal._utils.async_utils import synchronize_api
from modal._utils.auth_token_manager import _AuthTokenManager
from modal._utils.grpc_utils import DEFAULT_MAX_RETRIES
from modal.exception import AuthError, ExecutionError, PermissionDeniedError
from modal_proto import api_pb2

SECRET_KEY = "your-super-flexible-and-long-shared-secret-key"


@pytest.fixture
def auth_token_manager(client):
    """Create an AuthTokenManager instance for testing."""
    return _AuthTokenManager(client.stub)


@pytest.fixture
def valid_jwt_token():
    """Create a valid JWT token with expiry."""
    # Create a JWT with exp claim set to 1 hour from now
    exp = int(time.time()) + 3600
    payload = {"exp": exp, "type": "valid"}
    return jwt.encode(payload, SECRET_KEY, algorithm="HS256")


@pytest.fixture
def another_valid_jwt_token():
    """Create a valid JWT token with expiry."""
    # Create a JWT with exp claim set to 1 hour from now
    exp = int(time.time()) + 3600
    payload = {"exp": exp, "type": "another_valid"}
    return jwt.encode(payload, SECRET_KEY, algorithm="HS256")


@pytest.fixture
def expired_jwt_token():
    """Create an expired JWT token."""
    # Create a JWT with exp claim set to 1 hour ago
    exp = int(time.time()) - 3600
    payload = {"exp": exp, "type": "expired"}
    return jwt.encode(payload, SECRET_KEY, algorithm="HS256")


@pytest.fixture
def token_without_exp():
    """Create a JWT token without exp claim."""
    payload = {"type": "without_exp"}
    return jwt.encode(payload, SECRET_KEY, algorithm="HS256")


@pytest.fixture
def token_due_for_refresh():
    """Create a JWT token that is still valid but has used up its refresh fraction."""
    exp = int(time.time()) + 240  # 4 minutes from now
    payload = {"exp": exp, "type": "due_for_refresh"}
    return jwt.encode(payload, SECRET_KEY, algorithm="HS256")


@pytest.mark.asyncio
async def test_get_token_initial_fetch(auth_token_manager, valid_jwt_token, client, servicer):
    """Test getting token when no token exists."""

    # All these tests wrap get_token with @synchronize_api because they hang forever without it.
    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    servicer.auth_token = valid_jwt_token
    assert await wrapped_get_token.aio() == valid_jwt_token


@pytest.mark.asyncio
async def test_get_token_cached(auth_token_manager, valid_jwt_token, servicer):
    """Test that cached token is returned without making new request."""

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    # Set up initial token
    servicer.auth_token = valid_jwt_token
    assert await wrapped_get_token.aio() == valid_jwt_token

    # Set a bogus token in the servicer, and verify we get the cached valid token
    servicer.auth_token = "bogus"
    token = await wrapped_get_token.aio()
    assert token == valid_jwt_token


@pytest.mark.asyncio
async def test_get_token_expired(auth_token_manager, expired_jwt_token, valid_jwt_token, servicer):
    """Test that expired token triggers refresh."""
    # Set up expired token
    auth_token_manager._token = expired_jwt_token
    auth_token_manager._set_expiry(exp_time(expired_jwt_token))

    # Set up new token in servicer
    servicer.auth_token = valid_jwt_token

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    token = await wrapped_get_token.aio()
    assert token == valid_jwt_token
    assert auth_token_manager._token == valid_jwt_token


@pytest.mark.asyncio
async def test_get_token_needs_refresh(auth_token_manager, token_due_for_refresh, valid_jwt_token, servicer):
    """Test that a still-valid token is refreshed once it is past its refresh point."""
    auth_token_manager._token = token_due_for_refresh
    auth_token_manager._expiry = exp_time(token_due_for_refresh)
    auth_token_manager._refresh_at = time.time() - 1

    # Set up new token in servicer
    servicer.auth_token = valid_jwt_token

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    token = await wrapped_get_token.aio()
    assert token == valid_jwt_token
    assert auth_token_manager._token == valid_jwt_token


@pytest.mark.asyncio
async def test_get_token_expired_refresh_failure_falls_back(auth_token_manager, expired_jwt_token, monkeypatch):
    """Test that a failed refresh falls back to the expired cached token."""
    auth_token_manager._token = expired_jwt_token
    auth_token_manager._expiry = exp_time(expired_jwt_token)

    async def fail_auth_token_get(*args, **kwargs):
        raise RuntimeError("auth server unavailable")

    monkeypatch.setattr(auth_token_manager._stub, "AuthTokenGet", fail_auth_token_get)

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    assert await wrapped_get_token.aio() == expired_jwt_token


@pytest.mark.asyncio
async def test_get_token_without_cached_token_refresh_failure_raises(auth_token_manager, monkeypatch):
    """Test that a failed initial fetch still raises without a cached token."""

    async def fail_auth_token_get(*args, **kwargs):
        raise RuntimeError("auth server unavailable")

    monkeypatch.setattr(auth_token_manager._stub, "AuthTokenGet", fail_auth_token_get)

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    with pytest.raises(RuntimeError, match="auth server unavailable"):
        await wrapped_get_token.aio()


@pytest.mark.asyncio
async def test_get_token_fetch_only_retries_without_cached_token(auth_token_manager, valid_jwt_token, monkeypatch):
    """A failed fetch is only user-visible without a cached token, so that's the only case we retry."""
    captured: dict = {}

    async def recording_auth_token_get(request, **kwargs):
        captured.update(kwargs)
        return api_pb2.AuthTokenGetResponse(token=valid_jwt_token)

    monkeypatch.setattr(auth_token_manager._stub, "AuthTokenGet", recording_auth_token_get)

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    assert await wrapped_get_token.aio() == valid_jwt_token
    assert captured["retry"].max_retries == DEFAULT_MAX_RETRIES

    # The token is now cached, so a subsequent refresh falls back to it instead of retrying.
    auth_token_manager._set_expiry(0.0)
    assert await wrapped_get_token.aio() == valid_jwt_token
    assert captured["retry"].max_retries == 0


@pytest.mark.asyncio
async def test_get_token_expired_refresh_failure_backs_off(auth_token_manager, expired_jwt_token, monkeypatch):
    """After a failed refresh, the cached token is reused without re-hitting the server during the backoff window."""
    auth_token_manager._token = expired_jwt_token
    auth_token_manager._expiry = exp_time(expired_jwt_token)

    calls = 0

    async def fail_auth_token_get(*args, **kwargs):
        nonlocal calls
        calls += 1
        raise RuntimeError("auth server unavailable")

    monkeypatch.setattr(auth_token_manager._stub, "AuthTokenGet", fail_auth_token_get)

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    # First call attempts a refresh, fails, and falls back to the cached expired token.
    assert await wrapped_get_token.aio() == expired_jwt_token
    assert calls == 1
    assert auth_token_manager._in_backoff()
    expected_retry_after = time.time() + auth_token_manager.FAILURE_BACKOFF_BASE
    assert auth_token_manager._retry_after == pytest.approx(expected_retry_after, abs=1)

    # Second call is within the backoff window, so it returns the cached token without another RPC.
    assert await wrapped_get_token.aio() == expired_jwt_token
    assert calls == 1


@pytest.mark.asyncio
async def test_get_token_empty_response_backs_off(auth_token_manager, expired_jwt_token, monkeypatch):
    """An empty refresh response falls back to the cached token and enters backoff."""
    auth_token_manager._token = expired_jwt_token
    auth_token_manager._expiry = exp_time(expired_jwt_token)
    calls = 0

    async def empty_auth_token_get(*args, **kwargs):
        nonlocal calls
        calls += 1
        return api_pb2.AuthTokenGetResponse()

    monkeypatch.setattr(auth_token_manager._stub, "AuthTokenGet", empty_auth_token_get)

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    assert await wrapped_get_token.aio() == expired_jwt_token
    assert calls == 1
    assert auth_token_manager._in_backoff()
    assert await wrapped_get_token.aio() == expired_jwt_token
    assert calls == 1


@pytest.mark.asyncio
async def test_get_token_undecodable_response_backs_off(auth_token_manager, expired_jwt_token, monkeypatch):
    """An undecodable refresh response falls back to the cached token and enters backoff."""
    auth_token_manager._token = expired_jwt_token
    auth_token_manager._expiry = exp_time(expired_jwt_token)
    calls = 0

    async def undecodable_auth_token_get(*args, **kwargs):
        nonlocal calls
        calls += 1
        return api_pb2.AuthTokenGetResponse(token="not-a-jwt")

    monkeypatch.setattr(auth_token_manager._stub, "AuthTokenGet", undecodable_auth_token_get)

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    assert await wrapped_get_token.aio() == expired_jwt_token
    assert calls == 1
    assert auth_token_manager._in_backoff()
    assert await wrapped_get_token.aio() == expired_jwt_token
    assert calls == 1


@pytest.mark.asyncio
async def test_get_token_refresh_retried_after_backoff(
    auth_token_manager, expired_jwt_token, valid_jwt_token, servicer
):
    """Once the backoff window elapses, the next call refreshes again."""
    auth_token_manager._token = expired_jwt_token
    auth_token_manager._expiry = exp_time(expired_jwt_token)
    # Simulate an active backoff from a recent failure.
    auth_token_manager._retry_after = time.time() + auth_token_manager.FAILURE_BACKOFF_BASE
    servicer.auth_token = valid_jwt_token

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    # Within the backoff window: cached expired token is returned, no refresh.
    assert await wrapped_get_token.aio() == expired_jwt_token

    # Backoff has elapsed: the refresh happens and the new token is cached.
    auth_token_manager._retry_after = 0.0
    assert await wrapped_get_token.aio() == valid_jwt_token
    assert auth_token_manager._token == valid_jwt_token


@pytest.mark.parametrize("error_type", [AuthError, PermissionDeniedError])
@pytest.mark.asyncio
async def test_get_token_auth_denied_does_not_fall_back(auth_token_manager, expired_jwt_token, error_type, monkeypatch):
    """Auth-denial errors are surfaced instead of falling back to an expired cached token."""
    auth_token_manager._token = expired_jwt_token
    auth_token_manager._expiry = exp_time(expired_jwt_token)
    auth_token_manager._retry_after = 0.0
    auth_token_manager._backoff = auth_token_manager.FAILURE_BACKOFF_BASE
    error = error_type("credentials rejected")

    async def fail_auth_token_get(*args, **kwargs):
        raise error

    monkeypatch.setattr(auth_token_manager._stub, "AuthTokenGet", fail_auth_token_get)

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    with pytest.raises(error_type, match="credentials rejected"):
        await wrapped_get_token.aio()
    assert auth_token_manager._token == expired_jwt_token
    assert auth_token_manager._retry_after == 0.0
    assert auth_token_manager._backoff == auth_token_manager.FAILURE_BACKOFF_BASE
    assert not auth_token_manager._in_backoff()


@pytest.mark.asyncio
async def test_get_token_failure_backoff_grows_exponentially(
    auth_token_manager, expired_jwt_token, valid_jwt_token, monkeypatch
):
    """Consecutive refresh failures increase the cooldown, which resets after success."""
    now = 1_000.0
    monkeypatch.setattr(time, "time", lambda: now)
    auth_token_manager._token = expired_jwt_token
    auth_token_manager._set_expiry(0.0)
    should_fail = True

    async def auth_token_get(*args, **kwargs):
        if should_fail:
            raise RuntimeError("auth server unavailable")
        return api_pb2.AuthTokenGetResponse(token=valid_jwt_token)

    monkeypatch.setattr(auth_token_manager._stub, "AuthTokenGet", auth_token_get)

    expected_cooldowns = [0.5, 1.0, 2.0, 4.0, 8.0, 16.0, 32.0, 60.0, 60.0]
    for expected_cooldown in expected_cooldowns:
        auth_token_manager._retry_after = 0.0
        with pytest.raises(RuntimeError, match="auth server unavailable"):
            await auth_token_manager._refresh_token()
        assert auth_token_manager._retry_after == now + expected_cooldown

    should_fail = False
    auth_token_manager._retry_after = 0.0
    await auth_token_manager._refresh_token()
    assert auth_token_manager._backoff == auth_token_manager.FAILURE_BACKOFF_BASE

    should_fail = True
    auth_token_manager._set_expiry(0.0)
    auth_token_manager._retry_after = 0.0
    with pytest.raises(RuntimeError, match="auth server unavailable"):
        await auth_token_manager._refresh_token()
    assert auth_token_manager._retry_after == now + auth_token_manager.FAILURE_BACKOFF_BASE


@pytest.mark.asyncio
async def test_get_token_no_exp_claim(auth_token_manager, token_without_exp, servicer):
    """Test handling of token without exp claim."""
    servicer.auth_token = token_without_exp

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    token = await wrapped_get_token.aio()
    assert token == token_without_exp
    assert auth_token_manager._token == token_without_exp
    # Should use default expiry
    assert auth_token_manager._expiry > time.time()
    assert auth_token_manager._expiry <= time.time() + auth_token_manager.DEFAULT_EXPIRY_OFFSET
    # And should schedule the refresh partway through that lifetime.
    assert time.time() < auth_token_manager._refresh_at < auth_token_manager._expiry


@pytest.mark.asyncio
async def test_get_token_empty_response(auth_token_manager, servicer):
    """Test handling of empty token response."""
    servicer.auth_token = ""

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    with pytest.raises(ExecutionError):
        await wrapped_get_token.aio()


@pytest.mark.asyncio
async def test_get_token_none_response(auth_token_manager, servicer):
    """Test handling of None token response."""
    servicer.auth_token = None

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    with pytest.raises(ExecutionError):
        await wrapped_get_token.aio()


@pytest.mark.asyncio
async def test_concurrent_token_fetch(auth_token_manager, valid_jwt_token, servicer):
    """Test that concurrent calls don't make multiple requests."""
    servicer.auth_token = valid_jwt_token

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    # Make concurrent calls
    tasks = [wrapped_get_token.aio() for _ in range(5)]
    results = await asyncio.gather(*tasks)

    # All should return the same token
    assert all(token == valid_jwt_token for token in results)
    # The server should have been called only once.
    assert servicer.auth_tokens_generated == 1


@pytest.mark.asyncio
async def test_concurrent_refresh(auth_token_manager, token_due_for_refresh, valid_jwt_token, servicer):
    """Test that when get_token is called concurrently, test that old but valid token is returned."""
    # Set up token that needs refresh
    auth_token_manager._token = "old.but.valid.token"
    auth_token_manager._expiry = exp_time(token_due_for_refresh)
    auth_token_manager._refresh_at = time.time() - 1

    # Set up new token in servicer
    servicer.auth_token = valid_jwt_token
    servicer.auth_token_delay = 0.5

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    # Make concurrent calls
    tasks = [wrapped_get_token.aio() for _ in range(10)]
    results = await asyncio.gather(*tasks)

    # At least one call should have returned the new token
    assert valid_jwt_token in results
    # When called concurrently, only one coroutine should fetch a new token, and the others should use the older but
    # still valid token to improve throughput. Note, this isn't guaranteed, just very likely. May need to fix if flakey.
    assert "old.but.valid.token" in results
    # The new token should be cached
    assert auth_token_manager._token == valid_jwt_token


@pytest.mark.asyncio
async def test_concurrent_fetch_failure_makes_one_request(auth_token_manager, monkeypatch):
    """Concurrent callers with no cached token share one failing request instead of one request each."""
    calls = 0

    async def fail_auth_token_get(*args, **kwargs):
        nonlocal calls
        calls += 1
        await asyncio.sleep(0.1)
        raise RuntimeError("auth server unavailable")

    monkeypatch.setattr(auth_token_manager._stub, "AuthTokenGet", fail_auth_token_get)

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    results = await asyncio.gather(*[wrapped_get_token.aio() for _ in range(5)], return_exceptions=True)

    assert all(isinstance(result, RuntimeError) for result in results)
    assert calls == 1


@pytest.mark.asyncio
async def test_cancelling_a_caller_does_not_cancel_the_refresh(auth_token_manager, valid_jwt_token, servicer):
    """The refresh runs to completion for the remaining callers when the caller that started it is cancelled."""
    servicer.auth_token = valid_jwt_token
    servicer.auth_token_delay = 0.5

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    first = asyncio.create_task(wrapped_get_token.aio())
    await asyncio.sleep(0.1)
    second = asyncio.create_task(wrapped_get_token.aio())
    await asyncio.sleep(0.1)
    first.cancel()

    assert await second == valid_jwt_token
    assert servicer.auth_tokens_generated == 1


def test_decode_jwt_valid(valid_jwt_token):
    """Test JWT decoding with valid token."""
    decoded = _AuthTokenManager._decode_jwt(valid_jwt_token)
    assert "exp" in decoded
    assert "type" in decoded
    assert decoded["type"] == "valid"


def test_decode_jwt_without_exp(token_without_exp):
    """Test JWT decoding with token that has no exp claim."""
    decoded = _AuthTokenManager._decode_jwt(token_without_exp)
    assert "exp" not in decoded
    assert "type" in decoded
    assert decoded["type"] == "without_exp"


def test_decode_jwt_invalid_format():
    """Test JWT decoding with invalid token format."""
    with pytest.raises(ValueError):
        _AuthTokenManager._decode_jwt("invalid.token")


def test_needs_refresh_true(auth_token_manager):
    """Test _needs_refresh returns True once the refresh point has passed."""
    auth_token_manager._refresh_at = time.time() - 1
    assert auth_token_manager._needs_refresh() is True


def test_needs_refresh_false(auth_token_manager):
    """Test _needs_refresh returns False before the refresh point."""
    auth_token_manager._refresh_at = time.time() + 600
    assert auth_token_manager._needs_refresh() is False


def test_set_expiry_of_expired_token_refreshes_immediately(auth_token_manager):
    """Test an already-expired token is due for refresh rather than getting a negative delay."""
    auth_token_manager._set_expiry(time.time() - 100)
    assert auth_token_manager._needs_refresh() is True


def test_is_expired_true(auth_token_manager):
    """Test _is_expired returns True for expired token."""
    # Set expiry to 1 minute ago
    auth_token_manager._expiry = time.time() - 60
    assert auth_token_manager._is_expired() is True


def test_is_expired_false(auth_token_manager):
    """Test _is_expired returns False for valid token."""
    # Set expiry to 1 minute from now
    auth_token_manager._expiry = time.time() + 60
    assert auth_token_manager._is_expired() is False


@pytest.mark.asyncio
async def test_multiple_refresh_cycles(auth_token_manager, servicer):
    """Test multiple refresh cycles work correctly."""
    exp = int(time.time()) + 3600
    tokens = [
        jwt.encode({"exp": exp, "name": "t0"}, SECRET_KEY, algorithm="HS256"),
        jwt.encode({"exp": exp, "name": "t1"}, SECRET_KEY, algorithm="HS256"),
        jwt.encode({"exp": exp, "name": "t2"}, SECRET_KEY, algorithm="HS256"),
    ]

    @synchronize_api
    async def wrapped_get_token():
        return await auth_token_manager.get_token()

    # First call
    servicer.auth_token = tokens[0]
    token0 = await wrapped_get_token.aio()
    assert token0 == tokens[0]

    # Expire the token
    auth_token_manager._set_expiry(time.time() - 100)

    # Second call
    servicer.auth_token = tokens[1]
    token1 = await wrapped_get_token.aio()
    assert token1 == tokens[1]

    # Expire again
    auth_token_manager._set_expiry(time.time() - 100)

    # Third call
    servicer.auth_token = tokens[2]
    token2 = await wrapped_get_token.aio()
    assert token2 == tokens[2]


def exp_time(token: str):
    return jwt.decode(token, options={"verify_signature": False})["exp"]
